### PR TITLE
Proposal based votable state

### PR DESCRIPTION
**Pull request by mstrasinskis**
# Motivation

There are multiple reports from users who can’t vote. We believe the root cause of this issue is the 100-ballot limit for neurons. If a user has voted on more than 100 proposals (a likely scenario over a 4-day period), they don’t receive ballots for proposals they’ve already voted on, which remain technically votable due to this limit. The suggested solution is to switch from using neuron ballots to proposal ballots when calculating the votable state.

# Changes

- Use proposal ballots to get voting state.

# Tests

- Updated.

# Todos

- [ ] Add entry to changelog (if necessary).
Not necessary.

### CODE REVIEW

File: packages/nns/src/utils/neurons.utils.spec.ts [hunk 1]
```diff
@@ -135,40 +158,46 @@ describe("neurons-utils", () => {
...
+          {
+            neuronId: eligibleNeuronsDate[2].neuronId,
+            vote: Vote.Unspecified,
+            votingPower: BigInt(0),
+          },
+        ],
+      },
+      neurons: eligibleNeuronsDate,
     });
     expect(votable.length).toEqual(3);
```
Comment (dskloetd) [inline packages/nns/src/utils/neurons.utils.spec.ts:183]: same here?

File: packages/nns/src/utils/neurons.utils.spec.ts [hunk 2]
```diff
@@ -122,11 +137,19 @@ describe("neurons-utils", () => {
...
+          },
+        ],
+      },
       neurons: [
         {
           ...eligibleNeuronsDate[0],
-          recentBallots: [],
         },
       ],
     });
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts:170]: Extended.

File: packages/nns/src/utils/neurons.utils.spec.ts [hunk 3]
```diff
@@ -135,40 +158,46 @@ describe("neurons-utils", () => {
 
   it("should have votable neurons regardless of voting power", () => {
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts:174]: It’s probably from the time when neurons with a 0 balance were shown.

File: packages/nns/src/utils/neurons.utils.spec.ts (hunk 1 above)
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts]: Changed.

File: packages/nns/src/utils/neurons.utils.ts [hunk 4]
```diff
@@ -1,25 +1,24 @@
...
+import type { NeuronId } from "../types/common";
 import type {
   Ballot,
-  BallotInfo,
   NeuronInfo,
-  ProposalId,
   ProposalInfo,
 } from "../types/governance_converters";
 
 const voteForProposal = ({
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.ts]: Good point, renamed.

File: packages/nns/src/utils/neurons.utils.ts [hunk 5]
```diff
@@ -1,25 +1,24 @@
...
-  recentBallots,
-  proposalId,
+  ballots,
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.ts]: Done, it's cleaner.

File: packages/nns/src/utils/neurons.utils.spec.ts [hunk 6]
```diff
@@ -104,16 +116,19 @@ describe("neurons-utils", () => {
...
           ...eligibleNeuronsDate[0],
-          recentBallots: [
-            {
-              proposalId: BigInt(4),
-              vote: Vote.No,
-            },
-          ],
         },
       ],
     });
```
[the remark this answers is not in the document]
Comment (dskloetd) [reply packages/nns/src/utils/neurons.utils.spec.ts:144]: I still see 1 neuron (lines 139-143).
You've extended the ballots but not the neurons. But it's the neurons that are being filtered.

File: packages/nns/src/utils/neurons.utils.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dskloetd) [reply packages/nns/src/utils/neurons.utils.spec.ts:170]: I still see 1 neuron.

File: packages/nns/src/utils/neurons.utils.spec.ts [hunk 7]
```diff
@@ -162,25 +177,25 @@ describe("neurons-utils", () => {
...
             vote: Vote.Unspecified,
             votingPower: BigInt(0),
           },
         ],
       },
-      neurons: eligibleNeuronsDate,
+      neurons: eligibleNeuronsData,
     });
-    expect(votable.length).toEqual(3);
+    expect(votable.length).toEqual(2);
```
Comment (dskloetd) [inline packages/nns/src/utils/neurons.utils.spec.ts:198]: OK, here we start with 3 and end up with 2 neurons. So here filtering is being tested.
Maybe the purpose of the above tests isn't to test the filtering?
I'm just pointing it out but I leave it up to you to decide if the coverage is sufficient.

**Review by dskloetd — APPROVED** [review]

File: packages/nns/src/utils/neurons.utils.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts:144]: Correct, done.

File: packages/nns/src/utils/neurons.utils.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts:170]: Updated.

File: packages/nns/src/utils/neurons.utils.spec.ts (hunk 7 above)
Comment (mstrasinskis) [reply packages/nns/src/utils/neurons.utils.spec.ts]: These functions do the filtering, so it needs to be covered.

**Pull request merged**